### ISSUE TITLE
close of virtual keyboard in a mobile device browser triggers redundant search

### ISSUE THREAD

**Issue by shaun-madison** [running-coder/jquery-typeahead#225]
I have found the following snag that when viewing a page running jquery-typeahead in a mobile device browser (i.e.Chrome on an Android phone - specifically an S5) :

my TA plugin options are set for dynamic queries using an AJAX call - after typing in the query it sends the request and returns results, formats and opens the results list as expected BUT the virtual keyboard covers a good deal of the list - using the Android soft back button closes the keyboard but it triggers another send and refresh of the results even though the query has not been changed. 

I have also noticed this behaviour does NOT happen if your latest query was shorter than before, but always when it is longer or the initial query (the latest characters in the input field are underlined which seems to indicate a state where the input is new/dirty and closing the keyboard seems to generate an event that triggers the repeat lookup) 

Is this a glitch in the TA logic or is there a way to prevent this happening?

I have attached the javascript calling the TA functions for reference 

[ta_mcl_sitesearch_v2.js.txt](https://github.com/running-coder/jquery-typeahead/files/413320/ta_mcl_sitesearch_v2.js.txt)

**Comment by running-coder** [comment]
hey shaun thanks for reporting this! I'll have a look on how other search are handling the lack of space due to the keyboard pop'd up. Regarding an additional request being sent it's not too big of a deal since you should be getting the same results and the interface be unchanged but still annoying. The search is triggered on `input` event so I might have to sync on what is causing it on mobile device

**Comment by shaun-madison** [comment]
Hi Tom

Thanks for picking up on this issue. Having thought about it a bit more I have just realised the underlining of newly added characters in the query input field is actually due to having **predictive text set on.**

_When I change my phone settings to switch off predictive text and test what happens when you close the virtual keyboard using the soft back key I find it is no longer triggering phantom look-ups._

Unfortunately, at present, there seems to be a no reliable method to selectively disable predictive text on an HTML input field in a mobile browser. So we are a little closer to understanding the cause of the problem but not preventing.

**Comment by running-coder** [comment]
Perhaps https://davidwalsh.name/disable-autocorrect?

worth giving a try

**Comment by shaun-madison** [comment]
Hi Tom

I was also hoping the HTML5 autocomplete and autocorrect attributes this would be a solution but it seems from reading about them that: 
- they are not strictly adhered to by all mobile browsers
- they do not handle text prediction which is a feature of the mobile device's virtual keyboard 

Would it feasible to compare the last query string/filters and the current query string/filters when the input.typeahead event is triggered and check if there has been a real change to the input since the last input event and then cancel the event if not.

Shaun

**Comment by shaun-madison** [comment]
Hi Tom

I was also hoping the HTML5 autocomplete and autocorrect attributes would be a solution but it seems from reading about them that:

they are not strictly adhered to by all mobile browsers
they do not handle text prediction which is a feature of the mobile device's virtual keyboard
Would it be feasible to compare the last query string/filters and the current query string/filters when the input.typeahead event is triggered and check if there has been a real change to the input since the last input event and then cancel the event if not.

That would be similar to where you check a group' AJAX url to see if it is a duplicate of another group's url

Shaun

**Comment by shaun-madison** [comment]
Hi again

After considerable experimenting (in between watching the Olympics) I have just found this cute little fix for preventing a mobile phone, with predictive text enabled, from generating phantom lookups when you either close the virtual keyboard (with soft-back key) or when selecting an item in the results list (with a dynamic group):

```
$.typeahead( {

    input: ('#myITAinput'),     debug: true,     minLength: 3,

    maxItem: 0,     accent: true,     dynamic: true,

    delay: 500,     filter: false,},

    callback: {

		onSendRequest: function (node, query) {

			$("myITAinput").val('');            // clear input value (clears input dirtied by prediction)

			$("myITAinput").val(query);    // restore input with query - no more prediction hassles!

		},
   });
```

If you agree this simple fix does the job you may want to consider putting it into the main source code to avoid the problem for everyone.

Shaun

**Comment by running-coder** [comment]
hey shaun,

good stuff, It will work for your case where you set `minLength: 3` but I'm afraid it will have no effect on someone's `minLength: 0` to trigger an input on pageload for example. Perhaps there is some `event` property that is unique to open/close keyboard that could be validated, Ex:
```
if (event.someProperty) return false;
// Regular search
```

do you think you could print out the event generated by the open and also the close keyboard? I've searched the web a bit and couldn't find anything.

just add console.log(event) [here](https://github.com/running-coder/jquery-typeahead/blob/master/src/jquery.typeahead.js#L645)

**Comment by shaun-madison** [comment]
Hi Tom

From my observations I believe the behaviour is triggered by the predictive-text feature changing the input field's DOM state (in some way it is flagged as dirty / changed) NB: in some cases the text is not actually changed but just marked (characters underlined) as an active predictive text target. In this state when the focus is lost from the input field (closing the keyboard or clicking on a result list row) this seems to trigger another iteration of any dynamic group queries.

The behaviour only happens after you have again edited the input field content via the keyboard and not via JQuery access to the input field value. So I don't believe a  "generate on load" for a 0 minLength would have a problem - only after you have typed in another character - again deleting backwards only causes a single (as expected) lookup query to run (on my Sansung anyway)

I think one would have to use alerts rather than console logs as without an emulator I am not sure how you can access the browser console on a real mobile device?!

Shaun

**Comment by running-coder** [comment]
Hey shaun, I took some time to setup my iphone's w/ the computer to look at the console output and was unable to reproduce the issue. The search only occurs when a character is typed/removed and not when the keyboard is toggled. Perhaps it's happening for Android device but I'm unable to test on these. The issue is also minor as we can live with a data refresh without any negative effect to the user.

If you want to invest more time and find the issue, feel free to open a PR.

Thanks!
